File: src/peer/packetRouter.js
// Packet router — middleware chain for inbound data on PeerJS connections.
//
// Each middleware receives (remoteId, data, ctx) and returns true if it
// consumed the packet. The chain stops at the first handler that returns true.
//
// This replaces the nested if/switch inside useConnections.attachConn.onData.

import {
  isWireCiphertext,
  decryptWirePayload,
} from '../core/wireCrypto.js';
import {
  dispatchEphemeralInbound,
  dispatchReliableInbound,
} from '../messaging/messageProtocol.js';
import { hapticMessage } from '../core/haptics.js';
import { playSound } from '../core/sounds.js';
import { notifyNewMessage } from '../core/notifications.js';
import {
  getMessageById,
  saveMessage,
  updateMessage as dbUpdateMessagePayload,
  deleteMessageRow,
  saveVoiceBlob,
  deleteVoiceBlob,
  saveAvatar,
} from '../core/db.js';

// ─── Drop middleware ────────────────────────────────────────────

const DROP_TYPES = new Set([
  'drop-beacon', 'drop-beacon-ack', 'drop-req', 'drop-ack',
  'drop-rej', 'drop-cancel', 'file-start', 'file-chunk',
  'file-end', 'drop-resume',
]);

function isDropPacket(data) {
  return data && typeof data === 'object' && typeof data.type === 'string'
    && DROP_TYPES.has(data.type);
}

/**
 * Routes drop/file-transfer packets to DropManager.
 * Works on both ephemeral and reliable channels.
 */
export function dropMiddleware(remoteId, data, ctx) {
  if (!isDropPacket(data)) return false;
  ctx.dropHandlePacket?.(remoteId, data);
  return true;
}

// ─── Ephemeral channel middleware ────────────────────────────────

/**
 * Decrypts (if needed) and dispatches ephemeral-channel traffic
 * (typing indicators, heartbeats).
 */
export async function ephemeralMiddleware(remoteId, rawData, ctx) {
  let payload = rawData;

  if (isWireCiphertext(rawData)) {
    try { payload = await decryptWirePayload(remoteId, rawData); }
    catch (_) { return true; } // bad ciphertext — swallow
  } else if (typeof rawData === 'string') {
    return true; // ignore stray strings
  }

  if (!payload || typeof payload !== 'object') return true;

  // Drop packets on ephemeral channel
  if (isDropPacket(payload)) {
    ctx.dropHandlePacket?.(remoteId, payload);
    return true;
  }

  dispatchEphemeralInbound(payload, remoteId, {
    applyTyping: (isTyping) => ctx.applyTypingWithTimeout(remoteId, isTyping),
    onHeartbeat: () => ctx.onHeartbeat(remoteId),
  });
  return true;
}

// ─── Reliable channel middleware ────────────────────────────────

/**
 * Dispatches reliable-channel traffic (chat messages, wire handshake,
 * profile exchange, acks, etc).
 */
export async function reliableMiddleware(remoteId, data, ctx) {
  const handled = await dispatchReliableInbound(data, ctx.conn, remoteId, {
    localProfileRef: ctx.localProfileRef,
    peerIdRef: ctx.peerIdRef,
    seenMsgIdsRef: ctx.seenMsgIdsRef,
    setProfilesByPeer: ctx.setProfilesByPeer,
    saveAvatar,
    updateMessage: ctx.updateMessage,
    queueAckStatus: ctx.queueAckStatus,
    pushMessage: ctx.pushMessage,
    saveMessage,
    saveVoiceBlob,
    getMessageById,
    deleteMessageRow,
    deleteVoiceBlob,
    dbUpdateMessagePayload,
    setMessagesByPeer: ctx.setMessagesByPeer,
    notifyNewMessage,
    hapticMessage: () => {
      try {
        const raw = localStorage.getItem('orbits_chat_prefs_v1');
        const prefs = raw ? JSON.parse(raw) : {};
        if (prefs.vibration === false) return;
      } catch (_) {}
      hapticMessage();
    },
    playReceiveSound: () => playSound('receive'),
    sendEncrypted: (msg) => { void ctx.sendEncrypted(remoteId, msg); },
    onHandshakeError: (err) => { try { console.warn('[wire] handshake error', err); } catch (_) {} },
    onDecryptError: (err) => { try { console.warn('[wire] decrypt error', err); } catch (_) {} },
    onUnexpectedPlaintext: (d) => { try { console.warn('[wire] dropped unencrypted payload', d?.type || typeof d); } catch (_) {} },
  });

  if (handled) ctx.flushOutboxForPeer(remoteId);
  return true;
}

// ─── Chain runner ───────────────────────────────────────────────

/**
 * Build a packet handler for a given channel type.
 *
 * @param {'ephemeral'|'reliable'} channel
 * @param {string} remoteId
 * @param {object} ctx — handlers and refs from useConnections
 * @returns {(data: any) => Promise<void>}
 */
export function createPacketHandler(channel, remoteId, ctx) {
  if (channel === 'ephemeral') {
    return async (data) => {
      await ephemeralMiddleware(remoteId, data, ctx);
    };
  }

  // Reliable channel: drop packets first, then chat/wire traffic.
  const chain = [dropMiddleware, reliableMiddleware];

  return async (data) => {
    for (const mw of chain) {
      const consumed = await mw(remoteId, data, ctx);
      if (consumed) return;
    }
  };
}

File: lib/peer/packet_router.dart
// Port of `src/peer/packetRouter.js`.
//
// Middleware chain for inbound packets on a PeerJS DataConnection. Each
// middleware returns `true` when it consumed the packet; the chain stops
// at the first handler that returns `true`. The chain replaces the
// nested `if`/`switch` inside useConnections.attachConn.onData on the
// React side.
//
// Two channels, two chains:
//
//   ephemeral (unreliable UDP-like): [ephemeralMiddleware] — decrypts
//     wire ciphertext if needed, splits off drop-beacon packets, hands
//     the rest to [dispatchEphemeralInbound] (typing + heartbeat).
//
//   reliable (ordered TCP-like): [dropMiddleware] → [reliableMiddleware]
//     — drop/file-transfer traffic fast-paths to DropManager; everything
//     else (wire handshake, chat msg, profile exchange, bundle push/pull,
//     ack / edit / delete) goes through [dispatchReliableInbound].
//
// Notes relative to the JS version:
// - The giant inline callback bag in JS useConnections collapses into two
//   small Dart classes ([ReliableInboundCtx] + [EphemeralInboundCtx]) that
//   the caller builds once per peer at attach time. [PacketRouterCtx]
//   just composes them. Same wiring, less per-packet closure pressure.
// - `flushOutboxForPeer(remoteId)` in JS is a function of [remoteId];
//   here the caller bakes [remoteId] into [PacketRouterCtx.flushOutbox]
//   once at attach time because the Dart handler is also created per-peer.
// - DropManager is not ported yet — [PacketRouterCtx.dropHandlePacket] is
//   nullable so a dropless build still compiles and silently swallows
//   drop-beacon traffic (same effect as an unregistered JS handler).

import 'dart:async';

import '../core/wire_crypto.dart';
import '../messaging/message_protocol.dart';

/// Packet types that belong to the file-transfer (drop) subsystem. Match
/// the JS set 1:1 — adding or removing entries here will desync the two
/// builds. Source of truth: `src/peer/packetRouter.js::DROP_TYPES`.
const Set<String> dropTypes = <String>{
  'drop-beacon',
  'drop-beacon-ack',
  'drop-req',
  'drop-ack',
  'drop-rej',
  'drop-cancel',
  'file-start',
  'file-chunk',
  'file-end',
  'drop-resume',
};

/// True iff [data] is a JSON-shaped map whose `type` is one of [dropTypes].
/// Used by both the standalone [dropMiddleware] (reliable chain) and the
/// inline drop fast-path inside [ephemeralMiddleware].
bool isDropPacket(Object? data) {
  if (data is! Map) return false;
  final t = data['type'];
  return t is String && dropTypes.contains(t);
}

/// Signature for the handler returned by [createPacketHandler]. Matches
/// PeerJS `conn.on('data', ...)` — the transport gives us a raw decoded
/// value and we chew on it asynchronously.
typedef PacketHandler = Future<void> Function(Object? data);

/// Per-connection wiring for the router. Built once at attach time; the
/// router holds a reference for the lifetime of the channel.
class PacketRouterCtx {
  const PacketRouterCtx({
    required this.conn,
    required this.reliable,
    required this.ephemeral,
    required this.flushOutbox,
    this.dropHandlePacket,
  });

  /// Plaintext send over the underlying DataConnection. The router uses
  /// it to forward wire-handshake replies; app-level chat traffic flows
  /// through [reliable.sendEncrypted] instead.
  final ConnSend conn;

  /// Reliable-channel dispatch context for this peer.
  final ReliableInboundCtx reliable;

  /// Ephemeral-channel dispatch context (typing + heartbeat).
  final EphemeralInboundCtx ephemeral;

  /// Called after the reliable dispatcher consumes a packet — mirrors
  /// `flushOutboxForPeer(remoteId)` in JS. The caller typically retries
  /// any queued outbound messages here because an ack just landed.
  final void Function() flushOutbox;

  /// Optional DropManager hook. Null is a legal "dropless build" state —
  /// drop packets are silently discarded, which matches the JS behavior
  /// when `ctx.dropHandlePacket` is missing. Wired up in Phase 10+ once
  /// `drop_manager.dart` exists.
  final void Function(String remoteId, Map<String, Object?> data)?
      dropHandlePacket;
}

// ─── Middlewares ──────────────────────────────────────────────────

/// Reliable-channel drop-packet fast-path. Returns `true` if the packet
/// was a drop/file-transfer frame, `false` otherwise (fall through to
/// [reliableMiddleware]).
bool dropMiddleware(String remoteId, Object? data, PacketRouterCtx ctx) {
  if (!isDropPacket(data)) return false;
  ctx.dropHandlePacket
      ?.call(remoteId, Map<String, Object?>.from(data as Map));
  return true;
}

/// Ephemeral-channel entry. Decrypts wire ciphertext if needed, peels
/// off drop-beacon packets, and hands anything else to
/// [dispatchEphemeralInbound]. Always returns `true` — the ephemeral
/// chain has only this one middleware so nothing falls through.
Future<bool> ephemeralMiddleware(
  String remoteId,
  Object? rawData,
  PacketRouterCtx ctx,
) async {
  Object? payload = rawData;

  if (isWireCiphertext(rawData)) {
    try {
      payload = await decryptWirePayload(remoteId, rawData as String);
    } catch (_) {
      // Bad ciphertext on an unreliable channel is normal noise — swallow
      // without surfacing to onDecryptError (which is reliable-only in JS).
      return true;
    }
  } else if (rawData is String) {
    // Stray unencrypted string — ignore. Protocol puts typing/hb as JSON.
    return true;
  }

  if (payload is! Map) return true;
  final mapPayload = Map<String, Object?>.from(payload);

  // Drop-beacon packets can ride the ephemeral channel (lower latency for
  // discovery). Same short-circuit as the reliable drop middleware.
  if (isDropPacket(mapPayload)) {
    ctx.dropHandlePacket?.call(remoteId, mapPayload);
    return true;
  }

  dispatchEphemeralInbound(mapPayload, remoteId, ctx.ephemeral);
  return true;
}

/// Reliable-channel entry. Delegates to [dispatchReliableInbound] and
/// flushes the outbox when the dispatcher reports the packet was
/// consumed (handshake accepted, ack/message/etc routed). Always
/// returns `true` — this is the terminal middleware in the reliable
/// chain, so nothing falls through.
Future<bool> reliableMiddleware(
  String remoteId,
  Object? data,
  PacketRouterCtx ctx,
) async {
  final handled = await dispatchReliableInbound(
    data,
    ctx.conn,
    remoteId,
    ctx.reliable,
  );
  if (handled) {
    try {
      ctx.flushOutbox();
    } catch (_) {
      // Outbox flush is best-effort; nothing above this frame cares
      // whether it succeeded this tick.
    }
  }
  return true;
}

// ─── Chain runner ─────────────────────────────────────────────────

/// Build a packet handler for a given channel.
///
/// Pass `'ephemeral'` for the unreliable channel (typing + heartbeat +
/// optional drop-beacon) or `'reliable'` for the ordered channel (chat,
/// wire handshake, profile exchange, acks, drop/file-transfer frames).
///
/// The returned [PacketHandler] is safe to wire straight into the
/// transport's `onData` callback. It catches nothing — transport-level
/// failures should bubble so the connection layer can trigger a rekey
/// or a reconnect.
PacketHandler createPacketHandler(
  String channel,
  String remoteId,
  PacketRouterCtx ctx,
) {
  if (channel == 'ephemeral') {
    return (data) async {
      await ephemeralMiddleware(remoteId, data, ctx);
    };
  }

  // Reliable channel: drop packets first (synchronous fast-path), then
  // everything else.
  return (data) async {
    if (dropMiddleware(remoteId, data, ctx)) return;
    await reliableMiddleware(remoteId, data, ctx);
  };
}
